feat!: Network Management System v1

chichi13 opened the pull request:
## What is this PR about?

### Breaking Change

This PR introduces a complete redesign of network management in Dokploy, replacing the legacy `isolatedDeployment` feature with a flexible custom network system.

---

## Overview

This PR implements comprehensive Docker network management as a core feature of Dokploy. Users can now create, manage, and assign custom Docker networks to any resource type (applications, compose services, databases) with full control over network configuration, IPAM settings, and driver selection.

---

## Major Changes

### Network Management System

- **Global Network Management UI**: New `/dashboard/networks` page for organization-wide network administration
- **Resource-Level Network Assignment**: Integrated network management into all resource types (applications, compose, databases)
- **Project Organization**: Added dedicated networks tab to project views for better organization

### Database Schema

**New `network` Table**
- Multiple network drivers: `bridge` and `overlay`
- Advanced IPAM configuration: subnet, gateway, IP range
- Network encryption support for overlay networks
- Organization scope
- Docker network ID tracking for synchronization

**Resource Schema Updates**
- Added `customNetworkIds` field to all resource types:
  - `application`
  - `compose`
  - `postgres`, `mysql`, `mariadb`, `mongo`, `redis`
- Supports multiple network assignments per resource

### Network Features

#### Automatic Traefik Integration
- Traefik automatically connects to custom networks when domains are added to resources
- Automatic disconnection when networks are no longer in use by resources with domains
- Ensures proper routing without manual network configuration

#### Driver Validation
- Enforces overlay networks for Docker Swarm services
- Enforces overlay networks for compose stacks
- Bridge networks allowed for standard docker-compose mode

#### Network Synchronization (based on labels)
- Sync functionality to detect discrepancies between Docker and database state
- Identifies missing networks (in DB but not in Docker)
- Identifies orphaned networks (in Docker but not in DB)
- Import orphaned networks with proper label validation

### Migration and Backward Compatibility

#### Migration 0117: Add Network table and add `customNetworkIds` field

#### Migration 0118: Convert isolatedDeployment to customNetworkIds
This migration automatically converts all existing compose services using `isolatedDeployment=true` to the new network system:

1. Scans all compose services with `isolatedDeployment=true`
2. Creates network records using the compose `appName` as the network name
3. Assigns appropriate driver (`overlay` for remote servers, `bridge` for local)
4. Links networks to compose services via `customNetworkIds`
5. Sets `isolatedDeployment=false` on migrated services

**Context**: Compose services with isolated deployment already have Docker networks created with `name = compose.appName`. The migration imports these existing networks into the database and establishes proper tracking.

#### Migration 0119: Remove isolatedDeployment Field
Removes the obsolete `isolatedDeployment` boolean field from the database schema after successful migration.

---

## Testing

### New Test Coverage
- **Network Service Tests** (`network-service.test.ts`):
  - CRUD operations
  - Resource assignment/removal
  - Traefik integration
  - Sync and import functionality

- **Compose Network Integration** (`compose-networks.test.ts`):
  - Compose-specific network behavior
  - Stack vs docker-compose mode validation

- **Schema Validation** (`schema.test.ts`):
  - Network configuration validation
  - IPAM parameter validation
  - Driver validation

---

## Breaking Changes

### Removed: `isolatedDeployment` Field

The `isolatedDeployment` boolean field has been removed from the `compose` table schema.

**Migration**: All existing data is automatically migrated via migration 0118. No manual intervention required for Dokploy users.

## Checklist

Before submitting this PR, please make sure that:

- [x] You created a dedicated branch based on the `canary` branch.
- [x] You have read the suggestions in the CONTRIBUTING.md file https://github.com/Dokploy/dokploy/blob/canary/CONTRIBUTING.md#pull-request
- [x] You have tested this PR in your local instance.

## Issues related

closes #2798
closes #1495

## Screenshots

<img width="2554" height="1201" alt="image" src="https://github.com/user-attachments/assets/39971f25-d77c-4d78-ab51-03e0dca12a5e" />

<img width="2534" height="877" alt="image" src="https://github.com/user-attachments/assets/7c098e55-bbfa-4a88-8834-10551aef27d4" />

<img width="2257" height="555" alt="image" src="https://github.com/user-attachments/assets/0df798c7-a556-408c-9b93-bc462ca6395a" />

chichi13 committed:
feat(db): add network table for custom docker networks

chichi13 committed:
feat(db): add customNetworkIds field to all resources

chichi13 committed:
feat(services): add network service layer and docker utilities

chichi13 committed:
feat(api): add network tRPC router with resource filtering

chichi13 committed:
feat(docker): add custom network support to resource builders

chichi13 committed:
feat(traefik): auto-generate Docker service labels from domains

chichi13 committed:
feat(services): integrate network management in resource services

chichi13 committed:
feat(utils): add network support to backups and collision detection

chichi13 committed:
test: update tests to add customNetworkIds

chichi13 committed:
fix: format with biome

chichi13 committed:
feat(ui): add shared network management components

chichi13 committed:
feat(ui): add network management to applications

chichi13 committed:
feat(ui): add network management to compose services

chichi13 committed:
feat(ui): add network management to databases

chichi13 committed:
feat(ui): add global network management page and project tab

chichi13 committed:
feat(ui): integrate network management in navigation and pages

chichi13 committed:
fix(server): handle null port and convert uniqueConfigKey to string in Traefik labels

chichi13 committed:
feat(network): add Traefik auto-disconnect and improve code quality

chichi13 committed:
feat: add importOrphanedNetworks + avoid code duplication

chichi13 committed:
feat(db): backward compatibility with isolated network

chichi13 committed:
refactor: better network component + remove isolated network component

chichi13 committed:
feat(db)!: remove obsolete isolatedDeployment feature

BREAKING CHANGE: The isolatedDeployment feature has been completely removed from the codebase.

chichi13 committed:
test: add network tests

chichi13 committed:
feat(network): encrypted networks and remove attachable

chichi13 committed:
fix(network): stack compose without network should be attached to dokploy-network

chichi13 committed:
fix(network): network not updating for compose file

chichi13 committed:
fix: network can be deleted even if an application use it

chichi13 committed:
feat(network): better error management for network creation/deletion/update

chichi13 committed:
refactor(network): remove unused isDefault field from UI and API for network v1

chichi13 committed:
fix: tests

chichi13 committed:
feat(network): better error managerment + biome format

chichi13 committed:
style: better dark mode

chichi13 committed:
fix(network): network should be detached from Traefik if no more domain

chichi13 committed:
fix: traefik service use networkId not name

chichi13 committed:
feat(network): display internal flag in network selector

KarpachMarko commented:
Hi @chichi13. 
First of all thanks you for taking initiative on implementing this feature.
I started testing your PR from docker image you provided.
Could you please clarify if Traefik instance should automatically connect to newly created network, or that is what should be manually configured?

KarpachMarko commented:
Thanks for quick response. 
I tested again following your steps but same issue.
Here is what I did:
 - Created internal bridge network
 - Create new project with docker compose service
 - Assigned this internal network to created service
 - Set up domain name for the service
 - Redeployed the service
 
As a result no new file created in "Traefik File System"

Am I missing something?
Note: I have traefik running as standalone container. Not as docker service

chichi13 commented:
Indeed, will fix that, thanks!

Elvinra commented:
Hi,

I cannot choose where the network is created, i have tested with a remote servers, and the network is always created of the dokploy host, but my application are on a remote servers

chichi13 commented:
Hi,

I'll check that this afternoon, thanks!

chichi13 committed:
style: remove duplicate BreadcrumbSidebar on network page

chichi13 committed:
fix(network): deploy network on remote server

chichi13 committed:
feat(network): warn user when deploying domain on internal network

chichi13 commented:
@Elvinra is the network "db-prod" internal? If not Traefik will connect to both which can be improved.

But I don't really know how to do it:
- 3 networks attached
- 1 domain
- Which network Traefik should choose? Should I put a dropdown in the domain modal to select only one network?

Elvinra commented:
"db-prod" is not internal, dont know what internal should do ^^

Traefik have already a network attached "dokploy-network", why not use this one.
And attach "dokploy-network" to the service specified in the domain ?

chichi13 commented:
`internal`([internal](https://docs.docker.com/reference/cli/docker/network/create/#internal)) network are used essentially for container (like databases) which don't need internet access. Internal network won't have internet access + can't be published.

Attaching every application to `dokploy-network` is a security issue. You can check the issue #2798 if you want more details.

chichi13 committed:
feat: better domain management with network

chichi13 commented:
@Elvinra so, I've refactored how domain/traefik works, now you don't need to redeploy when you attach a domain. You can specify to which domain you want to attach. If you have 2 domains, both domains should be attached to the same network.

I would be happy to have a feedback from @Siumauricio to find out if there's a chance this PR will be merged (given the time I've spent on it, I'd really like it to be).

chichi13 committed:
feat(network): add preview network selector + previewNetworkIds in db

chichi13 committed:
fix: connect to preview selected network

Elvinra commented:
Hello,

I just pull the last version with a fresh docker, and i have an error

```
[
    {
        "error": {
            "json": {
                "message": "column projects_environments_applications_domains.networkId does not exist",
                "code": -32603,
                "data": {
                    "code": "INTERNAL_SERVER_ERROR",
                    "httpStatus": 500,
                    "path": "project.all",
                    "zodError": null
                }
            }
        }
    }
]
```

chichi13 committed:
fix: migration file

chichi13 commented:
I've updated the migration files + the docker image. I feel like I did the migration with drizzle incorrectly (I'm not very familiar with this tool).

chichi13 committed:
fix: traefik not disconnecting when no more domain attached to it

chichi13 committed:
feat: handle default network for domain in backend

chichi13 committed:
fix: disconnect traefik when editing

chichi13 commented:
Hello @Elvinra 

New migrations have been added to `canary` branch. So I'll need to merge canary to this branch and resolve conflict.

Thanks for testing and pointing out the regression. I'll try to resolve the problem this week.

A comment from @Siumauricio to get hist opinion for the PR would be great tbh

chichi13 committed:
Merge remote-tracking branch 'origin/canary' into feature/network-management

chichi13 committed:
fix: migration files

chichi13 commented:
@Elvinra Okay so I've resolved the conflicts.

> I cannot select dokploy-network in the domain modal, if i have any non public network attached to the application, pretty sure the problem exists on preview deployments.

I didn't quite understand when I replied the other day, but that's the expected behavior. The domain cannot be on `dokploy-network` since the application itself is only on the custom network (`Production` for example in my screen).

<img width="2173" height="338" alt="image" src="https://github.com/user-attachments/assets/5db0680d-2531-40fe-9e87-ca02ca8b20ca" />

<img width="923" height="338" alt="image" src="https://github.com/user-attachments/assets/d0050f25-398f-46c1-8c50-5ef4238f5c0d" />

You can only choose the networks (except `internal` networks) on which your app is located.

--- 

Image updated

Edgar-Alarcon commented:
Hi @Elvinra and @chichi13,

I'm following the progress of the Network Management PR, and since it introduces important improvements, I wanted to ask if there’s any update on its review or its chances of being approved.

Thanks for your time and for all the work on the project!

Siumauricio commented:
Can you resolve conflicts? @chichi13

chichi13 committed:
merge: integrate canary into feature/network-management

chichi13 committed:
Merge remote-tracking branch 'origin/canary' into feature/network-management

chichi13 committed:
fix: drizzle limitation to 100 columns per query

peteragurto commented:
> Okay, I understand. 
> 
> With the new migrations since the PR opened in October, some queries return more than 100 columns. Drizzle ORM limits to 100 columns, and we were at... 101. 
> 
> So I fixed the problem by specifying the columns. We've gone from 101 columns returned to ~7.

@Siumauricio I think it's time

autofix-ci[bot] committed:
[autofix.ci] apply automated fixes

Siumauricio committed:
chore: remove obsolete migration and related files for network management

- Deleted the SQL migration file `0145_needy_shooting_star.sql` which included the creation of the `network` table and related alterations.
- Updated the journal and snapshot metadata to reflect the removal of the migration.

Siumauricio committed:
Merge branch 'canary' into feature/network-management

Siumauricio committed:
feat: add network management schema and update related tables

- Introduced a new `network` table with relevant fields and constraints.
- Added `previewNetworkIds` and `customNetworkIds` columns to the `application`, `compose`, `mariadb`, `mongo`, `mysql`, `postgres`, and `redis` tables.
- Established foreign key relationships for the `network` and `domain` tables to ensure data integrity.

Siumauricio commented:
> @Siumauricio do something, delegate if you don't have time, too many PRs have been waiting for months. Wanting to control everything yourself goes against the spirit of the open source community...

I think we need to be more considerate of the time required for each feature. If a PR is not merged or reviewed, it is simply because it does not meet coding standards, especially for such a large feature. I am not going to merge a PR that does not meet the expectations of people who use dokploy. 

Unfortunately, AI moves so fast and sometimes we have too many PRs with AI. I'm not saying that's bad, it's just that the standard isn't being followed and this makes review very, very difficult. 

I have reviewed this PR and it definitely needs a lot of changes in terms of improving code quality (making it simpler, for example, or following the same structure that we follow throughout the project). That is why, if you see a PR that is not merged, it is simply because it does not follow our standards, and the best and easiest thing to do is to simply close the PR and create the functionality from scratch unfortunately. It saves time.

Siumauricio commented:
Hi @chichi13, thank you so much for this proposal and the thorough work on the network management redesign 🙏

We've now merged a network management system in #3774 that takes a different approach but shares the same core goals your PR set out — a flexible custom-network system and moving away from the legacy `isolatedDeployment` flag (which is now deprecated).

The implementation went through a significant restructuring: networks are assigned declaratively per-service (via the compose file / swarm spec) rather than through imperative `docker network connect`, which makes them survive restarts and Traefik recreations. It also enables cross-compose connectivity (e.g. sharing a database with another service through a dedicated network).

Because of that overlap I'll be closing this one, but your design input genuinely helped shape the direction. Thank you again for pushing this forward! 🚀
